server/api/roster/roster.controller.ts

```diff
@@ -409,37 +311,176 @@ class RosterController {
+        const needsQuotedValue = column.type === RosterColumnType.Date || column.type === RosterColumnType.DateTime;
+        const maybeQuote = (v: CustomColumnValue) => (v !== null && needsQuotedValue ? `'${v}'` : v);
+        const columnName = Roster.getColumnSelect(column);
+
+        if (op === 'between' || op === 'in') {
+          if (!Array.isArray(value)) {
+            throw new BadRequestError('Malformed search query. Expected array for value.');
+          }
+
+          if (op === 'in') {
+            return queryBuilder.andWhere(`${columnName} ${op} (:...${key})`, {
+              [key]: value,
+            });
+          }
+
+          if (op === 'between') {
+            return queryBuilder.andWhere(`${columnName} >= (:${key}Min) and ${columnName} <= (:${key}Max)`, {
+              [`${key}Min`]: maybeQuote(value[0]),
+              [`${key}Max`]: maybeQuote(value[1]),
+            });
+          }
+        }
+
+        if (Array.isArray(value)) {
+          throw new BadRequestError('Malformed search query. Expected scalar value.');
+        }
+
+        if (op === '=' || op === '<>' || op === '>' || op === '<') {
+          return queryBuilder.andWhere(`${columnName} ${op} :${key}`, {
+            [key]: maybeQuote(value),
+          });
+        }
 
-    res.json(updatedRosterEntry);
+        if (op === '~' || op === 'startsWith' || op === 'endsWith') {
+          const prefix = op !== 'startsWith' ? '%' : '';
+          const suffix = op !== 'endsWith' ? '%' : '';
+          return queryBuilder.andWhere(`${columnName} like :${key}`, {
+            [key]: `${prefix}${value}${suffix}`,
+          });
+        }
+
+        throw new BadRequestError('Malformed search query. Received unexpected value for "op".');
+      }, await Roster.queryAllowedRoster(org, role));
+  }
+
+  const queryBuilder = await makeQueryBuilder();
+
+  const rosterQuery = queryBuilder
+    .clone()
+    .offset(page * limit)
+    .limit(limit);
+  const sortColumn = rosterColumns.find(column => column.name === orderBy);
+  if (sortColumn) {
+    const order: OrderByCondition = {};
+    order[Roster.getColumnSelect(sortColumn)] = sortDirection;
+    rosterQuery.orderBy(order);
+  } else if (orderBy === 'unit') {
+    const order: OrderByCondition = {};
+    order['u.name'] = sortDirection;
+    rosterQuery.orderBy(order);
   }
+  const roster = await rosterQuery.getRawMany<RosterEntryData>();
 
+  const totalRowsCount = await queryBuilder.getCount();
+
+  return {
+    rows: roster,
+    totalRowsCount,
+  };
+}
+
+function copyRosterEntry(roster: Roster) {
+  const newEntry = new Roster();
+  newEntry.edipi = roster.edipi;
+  newEntry.unit = roster.unit;
+  newEntry.firstName = roster.firstName;
+  newEntry.lastName = roster.lastName;
+  newEntry.startDate = roster.startDate;
+  newEntry.endDate = roster.endDate;
+  newEntry.lastReported = roster.lastReported;
+  newEntry.customColumns = roster.customColumns;
```

Comment:
I think `roster.customColumns` needs to be deep cloned. Right now it looks like if the new entry's custom column values get modified it will also affect the original entry.